### PR TITLE
T/290: Tests are executed as a single entry point

### LINKED ISSUE #290: Check if tests would work better if we had one entry point in Karma
Right now we tell Karma to load all our test files, which is something around ~200 entry points. This is a lot and makes Webpack compile editor 200 times – so when running `npm t` we load 200 entry points.

When we've been setting up Karma I noticed that some guys recommend to have just one entry point, generated on-the-fly. This entry point should simply import all the tests which should be loaded.

It's worth checking if this could help cause our 7000 tests can only be run reliably in Chrome and Safari (https://github.com/ckeditor/ckeditor5/issues/573).

### PR DESCRIPTION
**Pull request by pomek**
### Suggested merge commit message ([convention](https://github.com/ckeditor/ckeditor5-design/wiki/Git-commit-message-convention))

Feature: Changed a way how the tests are executed. Karma will receive a single entry point which will import all tests. Closes #290.

### CODE REVIEW

File: packages/ckeditor5-dev-tests/lib/tasks/runautomatedtests.js [hunk 1]
```diff
@@ -16,7 +16,13 @@ module.exports = function runAutomatedTests( options ) {
 	return new Promise( ( resolve, reject ) => {
 		const config = getKarmaConfig( options );
 
+		if ( !config ) {
+			return reject();
```
Comment (ma2ciek) [inline packages/ckeditor5-dev-tests/lib/tasks/runautomatedtests.js]: Can we reject with an error?

File: packages/ckeditor5-dev-tests/lib/utils/automated-tests/getkarmaconfig.js [hunk 2]
```diff
@@ -6,67 +6,98 @@
...
 
-		if ( options.sourceMap ) {
-			preprocessorMap[ file ].push( 'sourcemap' );
+		if ( !files.length ) {
+			log.warn( 'Pattern "%s" does not match any file.', singlePattern );
 		}
+
+		allFiles.push( ...files.filter( file => {
+			// Ignore files which are saved in `manual/` directory. There are manual tests.
+			if ( file.match( /manual/ ) ) {
```
Comment (ma2ciek) [inline packages/ckeditor5-dev-tests/lib/utils/automated-tests/getkarmaconfig.js]: Maybe `tests[/\\]manual[/\\]` would be safer?

File: packages/ckeditor5-dev-tests/lib/utils/automated-tests/getkarmaconfig.js [hunk 3]
```diff
@@ -6,67 +6,98 @@
...
+				return false;
+			}
+
+			// Ignore `_utils` directory as well because there are saved utils for tests.
+			return !minimatch( file, path.join( '**', 'tests', '**', '_utils', '**', '*.js' ) );
```
Comment (ma2ciek) [inline packages/ckeditor5-dev-tests/lib/utils/automated-tests/getkarmaconfig.js]: The second path can be stored before the loop to not recreate it hundred of times.

File: packages/ckeditor5-dev-tests/lib/utils/automated-tests/getkarmaconfig.js [hunk 4]
```diff
@@ -6,67 +6,98 @@
...
+const AVAILABLE_REPORTERS = [
 	'mocha',
 	'dots'
 ];
 
-const coverageDir = path.join( process.cwd(), 'coverage' );
-
 /**
  * @param {Object} options
  * @returns {Object}
```
Comment (ma2ciek) [inline packages/ckeditor5-dev-tests/lib/utils/automated-tests/getkarmaconfig.js]: Or `undefined` if the error is logged in this function

File: packages/ckeditor5-dev-tests/lib/utils/automated-tests/getkarmaconfig.js [hunk 5]
```diff
@@ -6,67 +6,98 @@
...
  */
 module.exports = function getKarmaConfig( options ) {
+	const basePath = process.cwd();
+	const coverageDir = path.join( basePath, 'coverage' );
+	const log = karmaLogger.create( 'config' );
+
 	if ( !Array.isArray( options.files ) || options.files.length === 0 ) {
-		throw new Error( 'Karma requires files to tests. `options.files` has to be non-empty array.' );
+		return log.error( 'Karma requires files to tests. `options.files` has to be non-empty array.' );
```
Comment (ma2ciek) [inline packages/ckeditor5-dev-tests/lib/utils/automated-tests/getkarmaconfig.js]: Does Karma throw an error on `log.error`?

File: packages/ckeditor5-dev-tests/lib/utils/automated-tests/getkarmaconfig.js (hunk 2 above)
Comment (pomek) [reply packages/ckeditor5-dev-tests/lib/utils/automated-tests/getkarmaconfig.js]: No, it doesn't. Manual tests can be saved in subdirectories:

```
  '/ckeditor5-engine/tests/manual/tickets/887/1.js',
  '/ckeditor5-engine/tests/view/manual/clickobserver.js',
  '/ckeditor5-basic-styles/tests/manual/basic-styles.js'
```

But I replace the regexp with a glob pattern.

File: packages/ckeditor5-dev-tests/lib/utils/automated-tests/getkarmaconfig.js (hunk 4 above)
Comment (pomek) [reply packages/ckeditor5-dev-tests/lib/utils/automated-tests/getkarmaconfig.js]: I'm not sure whether we document it.

File: packages/ckeditor5-dev-tests/tests/utils/automated-tests/getkarmaconfig.js [hunk 6]
```diff
@@ -9,27 +9,60 @@ const mockery = require( 'mockery' );
...
+			tmp: {
+				fileSync: sandbox.stub().callsFake( () => stubs.tmp._fileSync ),
+				_fileSync: {
+					name: '',
+					removeCallback: sandbox.stub()
+				}
+			}
+		};
 
 		mockery.enable( {
```
Comment (ma2ciek) [inline packages/ckeditor5-dev-tests/tests/utils/automated-tests/getkarmaconfig.js]: Does it work with `useCleanCache` too? :D

Comment (pomek) [reply packages/ckeditor5-dev-tests/tests/utils/automated-tests/getkarmaconfig.js]: You can check :P

File: packages/ckeditor5-dev-tests/lib/utils/automated-tests/getkarmaconfig.js (hunk 4 above)
Comment (ma2ciek) [reply packages/ckeditor5-dev-tests/lib/utils/automated-tests/getkarmaconfig.js]: We document it if a function might return something else.

File: packages/ckeditor5-dev-tests/tests/utils/automated-tests/getkarmaconfig.js (hunk 6 above)
Comment (ma2ciek) [reply packages/ckeditor5-dev-tests/tests/utils/automated-tests/getkarmaconfig.js]: It works, I was actually surprised :D

**Pull request merged**